blog/content/post/2021-12-13-secure-jtag-on-imx-soc.md

```diff
@@ -0,0 +1,280 @@
+  - Security
+  - Manufacturing
+
+---
+## Introduction
+
+JTAG port is an important feature that helps a lot of engineers during product
+development. It also may be helpful for hackers while performing a
+reverse-engineering process - e.g. they can dump the memory of your firmware and
+get access to strictly confidential information. In this article, we show how to
+lock JTAG access for users who don't have a special key. We use for that Secure
+JTAG feature which is implemented on i.MX SoC's. In our lab, we tested this
+solution on i.MX6 but this same feature is available on a lot more NXP
+platforms. Unfortunately they have missing public documentation.
+
+## Requirements
+
+- **Development board**: Any devkit with i.MX6.
+
+- **SoC reference manual**: Setting JTAG in secure mode is a dangerous
+operation - if you miss something or will be wrong with any fusebit name or
+address you can enable unexcepted options. Here is an example: system is
+prepared to boot from SD card but you were wrong while fusemap understanding and
+now SoC always will be trying to boot from not existing eMMC memory. Read chip
+documentation and check twice that everything is correct.
+
+- **crucible tool**: JTAG mode and access key are saved in fuse bits - we need
+software to manipulate this in userspace.
+[crucible](https://github.com/f-secure-foundry/crucible/) tool can be used for
+that.
+
+- **JTAG compatible with OpenOCD**: On NXP sites we can read that Secure JTAG is
+supported only by the Lauterbach environment and ARM-DS5 IDE with DSTREAM
+debugger. In 3mdeb we are always trying to use open-source software.
+Unfortunately, OpenOCD doesn't officially support Secure JTAG in i.MX SoC's,
+but we found not accepted yet patch from 2014 which add support for work with
+SJC (System JTAG controller) in a secure mode. This patch work on OpenOCD from
+[official repository](git://git.code.sf.net/p/openocd/code) if you checkout this
+with `970a12aef` hash. There is a chance to apply this in the actual version of
+OpenOCD, but it needs some code changes. As a hardware, we used `ARM-USB-OCD-H`
+from Olimex.
+
+- **System image**: Any Linux image for your i.MX6 platform with kernel module
+`nvmem-imx-ocotp` for modify OTP memory.
+
+## How it works
+
+SoC's from i.MX family offers JTAG in three modes
+
+- **No debug** - security on the highest level, but after setting that
+development and debugging via JTAG port will be unavailable permanently on this
+device
+
+- **Secure JTAG** - mode described in this article, the best compromise between
+security and development possibilities - only person with access key can use
+JTAG
+
+- **Enabled JTAG** - default mode, enabled for anyone who has physical access to
+JTAG pins or even only device (is a chance to connect to the JTAG pins if the
+connector isn't soldered)
+
+Secure JTAG mode is based on challenge/response mechanism. SoC needs to save a
+unique challenge key during manufacturing. User can generate their response
```

Comment:
I'm not sure what means that SoCs need to save? Is it manufacturing action that is optional? Or maybe all SoC have challange key?

Reply:
All SoC have challenge key - I will rewrite this section to be more clear